controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -507,32 +514,296 @@ public void initializeStreamWriters(final EventStreamClientFactory clientFactory
+                streamMetadataStore.getStreamCutRecord(scope, stream, limits.getKey(), context, executor);
+        CompletableFuture<StreamCutRecord> limitMinFuture = limits.getValue() == null ? CompletableFuture.completedFuture(null) :
+                streamMetadataStore.getStreamCutRecord(scope, stream, limits.getValue(), context, executor);
+        return CompletableFuture.allOf(limitMaxFuture, limitMinFuture)
+                         .thenCompose(v -> {
+                             StreamCutRecord limitMax = limitMaxFuture.join();
+                             StreamCutRecord limitMin = limitMinFuture.join();
+                             if (limitMin != null) {
+                                 return streamMetadataStore.streamCutStrictlyGreaterThan(scope, stream, limitMin.getStreamCut(), lowerBound, context, executor)
+                                                    .thenCompose(gtMin -> {
+                                                        if (gtMin) {
+                                                            if (limitMax == null) {
+                                                                return CompletableFuture.completedFuture(lowerBound);
+                                                            } else {
+                                                                return streamMetadataStore.streamCutStrictlyGreaterThan(scope, stream, limitMax.getStreamCut(), lowerBound, context, executor)
+                                                                                          .thenApply(gtMax -> gtMax ? limitMax.getStreamCut() : lowerBound);
+                                                            }
+                                                        } else {
+                                                            return streamMetadataStore.streamCutStrictlyGreaterThan(scope, stream, lowerBound, limitMin.getStreamCut(), context, executor)
+                                                                               .thenApply(gt -> gt ? limitMin.getStreamCut() : null);
+                                                        }
+                                                    });
+                             } else {
+                                 // if min limit is 0 then latest effectively becomes the min. and we truncate at the lowerbound.
+                                 if (latest != null && policy.getConsumptionLimits().getMinValue() == 0L) {
+                                     // truncate at the lower bound 
+                                     return CompletableFuture.completedFuture(lowerBound);
+                                 } else {
+                                     return CompletableFuture.completedFuture(null);
+                                 }
+                             }
+                         });
+    }
+
+    private Map<Long, Long> computeSubscribersLowerBound(List<Map<StreamSegmentRecord, Long>> subscribers) {
```

Comment:
In the case where a Stream has only a single Subscriber, there is no need to compute a "lower bound" and we could short circuit this computation and directly return the Stream-Cut  corresponding to that Subscriber.

Reply:
It may be good to add a test case covering this scenario.

Reply:
created an issue to tackle it later. https://github.com/pravega/pravega/issues/5349

Reply:
It should be a small change. I think we should do it as part of this PR itself.

Reply:
the reason is "lowerbound" is Map<Long, Long> while subscriber's streamcut is Map<StreamSegmentRecord, Long> so there is one pass needed to convert even the single subscriber to Map<Long, Long>.
so the suggested optimization will not yield anything. 

what we can do instead is optimize the code a bit to not have to deal with these different types of maps but unify it.

Reply:
Ok. Got it.